proposals/0015-partitioned-epoch-reward-distribution.md

```diff
@@ -0,0 +1,362 @@
+the first block of the epoch boundary. This reduces the impact of rewards for
+vote account and also simplifies the overall changes. It also lets us focus on
+solving the primary bottleneck - Stake Rewards. Only Stake rewards are going to
+be distributed out over multiple blocks.
+
+In the new stake rewards distribution approach, we will separate the
+computation of rewards from the actual distribution of rewards at the epoch
+boundary by dividing the process into two distinct phases:
+
+   1. rewards calculation phase - during which the epoch rewards for all
+      activate stake accounts are computed and distribution chunks are scheduled.
+
+   1. rewards distribution phase - during which the calculated epoch rewards
+      for the active stake accounts are distributed.
+
+To help track and verify the reward distribution in the next phase, two new
+kinds of SysVar accounts are proposed : (a) `EpochRewardHistory` SysVar
+account and (b) `EpochRewardReserve` SysVar accounts. These SysVar accounts are
+updated at the boundary blocks (i.e. first and last blocks) in the reward
+calculation phase. More details of these two kinds of SysVar accounts are
+described in the following sections.
+
+### Rewards Calculation
+
+Reward calculation phase computes all the rewards that need to be distributed
+for the active stake accounts, and partitions the reward into a number of
+chunks for distribution in the next phase.
+
+Currently, on Solana Mainnet Beta with ~550K active stake accounts, it shows
+that epoch reward calculation takes around 10 seconds on average. This will
+make it impossible to perform rewards computation synchronous within one block.
+To make this work, a design of complete async reward computation is required.
+
+However, there are quite a few promising optimizations that can cut down the
+reward computation time. An experiment for reward calculation optimization
+(https://github.com/solana-labs/solana/pull/31193) showed that we can achieve
+the reward calculation time down to **40ms**. This is well below the 400ms
+block time.
+
+Therefore, the following of the design is based on the above optimization. The
+reward calculation will be performed at the first block of the epoch. The full
+rewards result is calculated, the result is partitioned into distribution
+chunks stored in the bank, which will then be distributed during the `reward
+distribution` phase.
+
+To ensure that each block distributes a subset of the rewards in a
+deterministic manner for the current epoch, while also randomizing the
+distribution across different epochs, the partitioning of all rewards will be
+done as follows.
+
+First, the reward results are sorted by account's Pubkey, and randomly shuffled
+with a fast `rng` seeded by current epoch. The shuffled reward results are then
+divided into `M` chunks. This process will ensure that the reward distribution
+is deterministic for the current epoch, while also introducing a degree of
+randomness between epochs.
```

Comment:
Yes, it will.

No, the metric didn't have that. But @t-nelson have benchmarked. It seems to only take around 7ms.

https://discord.com/channels/428295358100013066/960593861732884520/1096589073050783764

Reply:
Yeah. That will be the case if we decided to introduce the reserve sysvar. But whether to have these sysvar is still an open question.  If we choose not to have those sysvars, then we can avoid those stores.